File: backend/app/schemas/transaction.py
from pydantic import BaseModel
from typing import Optional
from datetime import date, datetime


class TransactionOut(BaseModel):
    id: int
    statement_id: int
    txn_date: date
    ref_number: Optional[str]
    description: str
    merchant_name: Optional[str]
    amount: float
    txn_type: str
    balance_after: Optional[float]
    category_id: Optional[int]
    is_categorized: bool
    created_at: datetime

    class Config:
        from_attributes = True


class CategoryPatch(BaseModel):
    category_id: int


class BulkCategorize(BaseModel):
    transaction_ids: list[int]
    category_id: int

File: backend-java/src/main/java/com/expensetracker/model/Transaction.java
package com.expensetracker.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.OffsetDateTime;

@Entity
@Table(name = "transactions")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(onlyExplicitlyIncluded = true)
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @EqualsAndHashCode.Include
    private Long id;

    @Column(name = "user_id", nullable = false)
    private Long userId;

    @Column(name = "statement_id", nullable = false)
    private Long statementId;

    @Column(name = "txn_date", nullable = false)
    private LocalDate txnDate;

    @Column(name = "ref_number")
    private String refNumber;

    @Column(nullable = false)
    private String description;

    @Column(name = "merchant_name")
    private String merchantName;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal amount;

    @Column(name = "txn_type", nullable = false)
    private String txnType;

    @Column(name = "balance_after", precision = 12, scale = 2)
    private BigDecimal balanceAfter;

    @Column(name = "category_id")
    private Long categoryId;

    @Column(name = "is_categorized", nullable = false)
    @Builder.Default
    private boolean isCategorized = false;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;
}
